lib/improver/utilities/cube_checker.py

```diff
@@ -35,47 +35,92 @@
 import numpy as np
 
 
-def check_cube_not_float64(cube, fix=False):
-    """Check a cube does not contain any float64 data, excepting time
-    coordinates. The cube can be modified in place, if the fix keyword is
+def check_cube_datatypes(cube, fix=False):
+    """Check the data within a cube is float64.
+    Coordinates are also checked to ensure that they have the desired datatype.
+    The cube can be modified in place, if the fix keyword is
     specified to be True.
 
     Args:
         cube (iris.cube.Cube):
-            The input cube that will be checked for float64 inclusion.
+            The input cube that will be checked for whether it is float64.
 
     Keyword Args:
         fix (bool):
-            If fix is True, then the cube is amended to not include float64
-            data, otherwise, an error will be raised if float64 data is found.
+            If fix is True, then the cube is amended to not include erroneous
+            datatypes, otherwise, an error will be raised.
 
     Raises:
-        TypeError : Raised if float64 values are found in the cube.
+        TypeError : Raised if the cube is not of the datatype specified, and
+            'fix' is not set to True.
 
     """
     if cube.dtype == np.float64:
         if fix:
             cube.data = cube.data.astype(np.float32)
         else:
-            raise TypeError("64 bit cube not allowed: {!r}".format(cube))
+            msg = ("The {} cube is expected to be of float32 datatype. "
+                   "The cube provided was of float64 datatype.".format(
+                       cube.name()))
+            raise TypeError(msg)
     for coord in cube.coords():
         if coord.name() in ["time", "forecast_reference_time"]:
-            continue
-        if coord.points.dtype == np.float64:
-            if fix:
-                coord.points = coord.points.astype(np.float32)
-            else:
-                raise TypeError(
-                    "64 bit coord points not allowed: {} in {!r}".format(
-                        coord, cube))
-        if (hasattr(coord, "bounds") and coord.bounds is not None and
-                coord.bounds.dtype == np.float64):
-            if fix:
-                coord.bounds = coord.bounds.astype(np.float32)
-            else:
-                raise TypeError(
-                    "64 bit coord bounds not allowed: {} in {!r}".format(
-                        coord, cube))
+            check_coord_datatypes(coord, np.int64, fix=True, rounding=True)
+        elif coord.name() in ["forecast_period", "realization"]:
+            check_coord_datatypes(coord, np.int32, fix=True, rounding=False)
+        else:
+            check_coord_datatypes(coord, np.float32, fix=True, rounding=False)
```

Comment:
These statements all force the time coordinates to integers without checking the units.  If for example the coordinates are provided in hours, forcing the value to integer (by rounding) in some cases (eg 15 minute cycling nowcasts) will change the value to something incorrect.

Some options that might be more desirable:
- Change the `fix` value in these calls to match that being passed into the function (ie only force the change when explicitly called / option to fail not fix)
- Include conversion to units of "seconds" in this function to ensure no values will be incorrectly changed